File: 07 No Space Left On Device/no_space_left_on_device.py
# https://adventofcode.com/2022/day/7

import heapq
from dataclasses import dataclass, field

input_file = 'real_input.txt'
#input_file = 'test_input.txt'

with open(input_file) as file:
    terminal_output = [line.split() for line in file.readlines()]


@dataclass
class Dir:
    name: str
    parent: None = None
    subdirs: dict = field(default_factory=dict)
    files: dict = field(default_factory=dict)


def build_tree(terminal_output):
    root = Dir('/')
    curr_dir = None
    ls = False
    for line in terminal_output:
        if line[0] == '$':
            ls = False
            if line[1] == 'cd':
                if line[2] == '/':
                    curr_dir = root
                elif line[2] == '..':
                    curr_dir = curr_dir.parent
                else:
                    curr_dir = curr_dir.subdirs[line[2]]
            elif line[1] == 'ls':
                ls = True
        elif ls:
            if line[0] == 'dir':
                curr_dir.subdirs[line[1]] = Dir(line[1], curr_dir)
            else:
                curr_dir.files[line[1]] = int(line[0])
    return root


def folder_sizes(fs):
    sizes_ascending = []
    def size(curr_dir):
        file_sizes = sum(curr_dir.files.values())
        subdir_sizes = sum(size(subdir) for subdir in curr_dir.subdirs.values())
        total_size = file_sizes + subdir_sizes
        heapq.heappush(sizes_ascending, total_size)
        return total_size
    root_size = size(fs)
    return sizes_ascending, root_size


# Part 1

fs = build_tree(terminal_output)
sizes_sorted, root_size = folder_sizes(fs)

sizes = sizes_sorted.copy()
sum_under_100k = 0
while True:
    curr = heapq.heappop(sizes)
    if curr <= 100000:
        sum_under_100k += curr
    else:
        break
print(sum_under_100k)


# Part 2

DISK_SPACE = 70000000
SPACE_REQUIRED = 30000000

sizes = sizes_sorted.copy()
while True:
    size = heapq.heappop(sizes)
    if size + (DISK_SPACE - root_size) >= SPACE_REQUIRED:
        print(size)
        break

File: 07 No Space Left On Device/no_space_left_on_device.js
// https://adventofcode.com/2022/day/6

const fs = require("fs");

let input_file = "real_input.txt";
//input_file = "test_input.txt";

const terminal_output = fs.readFileSync(input_file, "utf8").split("\n");

class Dir {
    constructor(name, parent = null) {
        this.name = name;
        this.parent = parent;
        this.subdirs = {};
        this.files = {};
    }
}

function build_tree(terminal_output) {
    const root = new Dir("/");
    let curr_dir = null;
    let ls = false;
    for (let i in terminal_output) {
        let line = terminal_output[i].trimEnd().split(" ");
        if (line[0] === "$") {
            ls = false;
            if (line[1] === "cd") {
                if (line[2][0] === "/") {
                    curr_dir = root;
                } else if (line[2] === "..") {
                    curr_dir = curr_dir.parent;
                } else {
                    curr_dir = curr_dir.subdirs[line[2]];
                }
            } else if (line[1] === "ls") {
                ls = true;
            }
        } else if (!!ls) {
            if (line[0] === "dir") {
                curr_dir.subdirs[line[1]] = new Dir(line[1], curr_dir);
            } else {
                curr_dir.files[line[1]] = Number(line[0]);
            }
        }
    }
    return root;
}

function folder_sizes(tree) {
    let walk_down = [tree];
    let walk_up = [];
    // Walk down
    while (walk_down.length > 0) {
        let curr_dir = walk_down.pop();
        for (let i in curr_dir.subdirs) {
            walk_down.push(curr_dir.subdirs[i]);
        }
        walk_up.push(curr_dir);
    }
    let sizes_map = new Map();
    let sizes_arr = [];
    // Walk up
    while (walk_up.length > 0) {
        let curr_dir = walk_up.pop();
        let file_sizes = 0;
        for (let i in curr_dir.files) {
            file_sizes += curr_dir.files[i];
        }
        let subdir_sizes = 0;
        for (let i in curr_dir.subdirs) {
            subdir_sizes += sizes_map.get(curr_dir.subdirs[i]);
        }
        let total_size = file_sizes + subdir_sizes;
        sizes_map.set(curr_dir, total_size);
        sizes_arr.push(total_size);
    }
    return sizes_arr;
}

const tree = build_tree(terminal_output);
const sizes = folder_sizes(tree);


// Part 1

let sum_sizes = 0;
for (let size of sizes) {
    if (size <= 100000) sum_sizes += size;
}
console.log(sum_sizes);


// Part 2

const DISK_SPACE = 70000000;
const SPACE_REQUIRED = 30000000;

sizes.sort((a, b) => a - b);
for (let size of sizes) {
    if (size + (DISK_SPACE - sizes[sizes.length - 1]) >= SPACE_REQUIRED) {
        console.log(size);
        break;
    }
}
